Add option to generate ESM exports instead of CJS (#1523)

rehanvdm opened the pull request:
**What issue does this pull request resolve?**
Adds an option so that the generated code can do both CJS and ESM exports. https://github.com/ajv-validator/ajv/issues/1523

**What changes did you make?**
Adde an extra flag  `code.exportEsm` to generate code with ESM exports instead of CJS.

**Is there anything that requires more attention while reviewing?**
I wanted to use the `code.es5` flag and then make the assumption that ES5 code uses CJS and ES6 uses ESM. But this would have been a breaking change because by default the code is exporting as ES6. So I thought it best to add an additional flag called `code.exportEsm`. It defaults to false so it introduces no breaking changes.

CJS (default) creates exports as `module.exports = ` for single exports and `exports.` or `exports[` for multiple exports. ESM creates exports as `export const` and  `export default` for single exports, where the `export const` has been given a fixed name of `validate`. This is just to cater for both import methods. Then for the multiple exports, the `export default` option is not available(or has not been implemented) and must be done with named exports as in `export const`.

I tested both scenarios for when the code has a single and multiple exports. Note that ESM can not export with special characters in the name, CJS can because it then uses the array syntax like `exports[https://example.com/string.json]`. I didn't just want it to fail the ESM version when the CJS passed so instead I used some regex to replace special characters and then the ESM export would look like `export const https___example_com_string_json`

I also needed to add a new devDependencies called `module-from-string` to be able to test the generated code. It is similar to `require-from-string`. The [package](https://www.npmjs.com/package/module-from-string) is almost a year old with not a lot of stars or downloads but looks stable and was good enough for the tests. I am not sure how else to test ESM modules, open to suggestions if you do not like this method.

Commit:
feat: add option to generate ESM exports instead of CJS (#1523)

rehanvdm committed:
Merge branch 'master' into standalone-generate-esm

epoberezkin commented on docs/options.md:
```diff
@@ -65,6 +65,7 @@ const defaultOptions = {
   code: {
     // NEW
     es5: false,
+    exportEsm: false,
```
maybe just esm?

also I was wondering, it’s not a problem if both es5 and esm are combined, esm export can be limited to es5 syntax too, so they are indeed orthogonal…

rehanvdm replied:
>maybe just esm? 

That makes sense, follows the naming convention of the es5 option

>also I was wondering, it’s not a problem if both es5 and esm are combined, esm export can be limited to es5 syntax too, so they are indeed orthogonal…

Correct me if I am wrong, but at default es5 is false, so it uses es6. If we assume that many are using the default, it would be a breaking change if we make the rule that es6 exports in ESM. That is what happened while I was developing, many of the tests failed because they didn't specify es5 true, meaning es6 + ESM and then the tests failed because the moduleCode was still using `requireFromString` which did not work because the code is now a module so I had to use `importFromStringSync`. That is why I don't think they can be combined?

epoberezkin replied:
Sorry, by “combined” I meant that it will be ok if a user uses both options. These should be separate options in ajv, you are right.

rehanvdm replied:
Okay cool

epoberezkin commented on lib/compile/codegen/code.ts:
```diff
@@ -155,6 +155,19 @@ export function getProperty(key: Code | string | number): Code {
 }
 
+//Does best effort to format the name properly
+export function getEsmExportName(key: Code | string | number): Code {
```
To be honest, I think it’s better to simply fail if user attempts to use a non-identifier as schema key, then try to adjust based on some rules that a user does not know producing unexpected names.

rehanvdm replied:
Makes sense, just though I do it this way as I saw ES Build do it like this. Will change it to throw an error. Any idea for a good error message? I saw in the tests that it is possible to alias $ids 

```js
     const moduleCode = standaloneCode(ajv, {
          validateNumber: "https://example.com/number.json",
          validateString: "https://example.com/string.json",
        })
```

Maybe mention the solution like that in the error?

epoberezkin replied:
Something like: invalid export name ${name}. You can use an explicit name mapping.

Do the docs not have this name mapping, is it only in tests?

rehanvdm replied:
I was about to say no it doesn't, but after about 15minutes of reading the docs I see that it is mentioned only in code comments here: https://ajv.js.org/standalone.html#usage-from-code. When I initially read it last week when I started to really dig into AJV, I just didn't notice it and did not understand why it is needed so thought it isn't important. 

It might be good to somehow highlight that if `code.esm` is true then your ids must be valid/exportable names and that you can remap them like that? Otherwise you will get that error.

epoberezkin replied:
yes, worth adding to the docs, but can be a separate PR

epoberezkin commented on lib/standalone/index.ts:
```diff
@@ -43,7 +47,15 @@ function standaloneCode(
       if (v) {
         const vCode = validateCode(usedValues, v.source)
-        code = _`${code}${_n}exports${getProperty(name)} = ${v.source?.validateName};${_n}${vCode}`
+        if (ajv.opts.code.exportEsm) {
```
I am tempted to refactor common parts out

rehanvdm replied:
Okay will do for this IF block or did you have something bigger in mind?

epoberezkin replied:
No, only this if

epoberezkin commented on docs/options.md:
```diff
@@ -65,6 +65,7 @@ const defaultOptions = {
   code: {
     // NEW
     es5: false,
+    exportEsm: false,
```
```suggestion
    esm: false,
```

epoberezkin commented on docs/options.md:
````diff
@@ -347,6 +348,9 @@ Code generation options:
 ```typescript
 type CodeOptions = {
   es5?: boolean // to generate es5 code - by default code is es6, with "for-of" loops, "let" and "const"
+  exportEsm?: boolean // how functions should be exported - by default CJS is used, so the validate function(s) 
````
```suggestion
  esm?: boolean // how functions should be exported - by default CJS is used, so the validate function(s) 
```

epoberezkin commented on lib/core.ts:
```diff
@@ -140,6 +140,7 @@ export interface CurrentOptions {
 
 export interface CodeOptions {
   es5?: boolean
+  exportEsm?: boolean
```
```suggestion
  esm?: boolean
```

epoberezkin commented on lib/standalone/index.ts:
```diff
@@ -43,7 +47,15 @@ function standaloneCode(
       if (v) {
         const vCode = validateCode(usedValues, v.source)
-        code = _`${code}${_n}exports${getProperty(name)} = ${v.source?.validateName};${_n}${vCode}`
+        if (ajv.opts.code.exportEsm) {
```
```suggestion
        if (ajv.opts.code.esm) {
```

epoberezkin commented on spec/standalone.spec.ts:
```diff
@@ -21,31 +42,61 @@ describe("standalone code generation", () => {
 
-      it("should generate module code with named exports", () => {
+      it("should generate module code with named export - ESM", () => {
+        ajv = new _Ajv({code: {source: true, exportEsm: true}})
```
```suggestion
        ajv = new _Ajv({code: {source: true, esm: true}})
```

epoberezkin commented on spec/standalone.spec.ts:
```diff
@@ -21,31 +42,61 @@ describe("standalone code generation", () => {
       })
+
+      it("should generate module code with all exports - ESM", () => {
+        ajv = new _Ajv({code: {source: true, exportEsm: true}})
```
```suggestion
        ajv = new _Ajv({code: {source: true, esm: true}})
```

epoberezkin commented on spec/standalone.spec.ts:
```diff
@@ -242,6 +294,26 @@ describe("standalone code generation", () => {
   })
 
+  it("should generate module code with a single export - ESM", () => {
+    const ajv = new _Ajv({code: {source: true, exportEsm: true}})
```
```suggestion
    const ajv = new _Ajv({code: {source: true, esm: true}})
```

Commit:
feat: add option to generate ESM exports instead of CJS (#1523)
- Renamed exportEsm to esm
- Extracted common code
- Changed invalid export names to rather throw an error

epoberezkin commented on lib/standalone/index.ts:
```diff
@@ -43,7 +47,10 @@ function standaloneCode(
       if (v) {
         const vCode = validateCode(usedValues, v.source)
-        code = _`${code}${_n}exports${getProperty(name)} = ${v.source?.validateName};${_n}${vCode}`
+        const exportSyntax = ajv.opts.code.esm
```
cool
